pkg/visitors/assemble.go

```diff
@@ -71,6 +72,86 @@ func (v *Assemble) Visit(f uefi.Firmware) error {
 		}
 
 		fileOffset := f.DataOffset
+
+		var pad *uefi.File
+		var shrink uint64
+		for _, file := range f.Files {
+			fileBuf := file.Buf()
+			fileLen := uint64(len(fileBuf))
+			if fileLen == 0 {
+				log.Fatal("Unexpected empty file buffer (at least a header was expected): %v", file.Header.GUID)
+			}
+			if file.Header.GUID == *uefi.FFGUID && fileLen > 24 {
+				// found non-empty PAD to check for free space.
+				pad = file
+			}
+
+			// Pad to the 8 byte alignments.
+			alignedOffset := uefi.Align8(fileOffset)
+			// Read out the file alignment requirements
```

Comment:
Please don't copy this code from further below, if you need to do a two pass, lets talk about moving it out into a helper function.

Reply:
I'm not sure how that would work.  Are you saying just try to do a normal insert and then on failure try the use the pad file space?  My approach was to look first if there is a space problem and then if there is, do the pad file method since while figuring out the space problem, it gives how much the pad file needs to be reduced.  If there is no pad file space then error out.  If there isn't a space problem then do the normal insert.